crml/cennzx-spot/src/types.rs

```diff
@@ -0,0 +1,158 @@
+// Copyright (C) 2019 Centrality Investments Limited
+// This file is part of CENNZnet.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+//!
+//! CENNZX-SPOT Types
+//!
+use core::convert::TryInto;
+use uint::construct_uint;
+construct_uint! {
+	/// 128-bit unsigned integer.
+	pub struct U128(2);
+}
+
+construct_uint! {
+	/// 256-bit unsigned integer.
+	pub struct U256(4);
+}
+
+use codec::{Compact, CompactAs, Decode, Encode};
+
+/// FeeRate S.F precision
+const SCALE_FACTOR: u128 = 1_000_000;
+
+/// FeeRate (based on Permill), uses a scale factor
+/// Inner type is `u128` in order to support compatibility with `generic_asset::Balance` type
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[derive(Encode, Decode, Default, Copy, Clone, PartialEq, Eq, Debug)]
+pub struct FeeRate(u128);
+
+impl FeeRate {
+	/// Create FeeRate as a decimal where `x / 1000`
+	pub fn from_milli(x: u128) -> FeeRate {
+		FeeRate(x * SCALE_FACTOR / 1000)
+	}
+
+	/// Create a FeeRate from a % i.e. `x / 100`
+	pub fn from_percent(x: u128) -> FeeRate {
+		FeeRate(x * SCALE_FACTOR / 100)
+	}
+
+	/// Divide a `TryInto<u128>` supported numeric by a FeeRate
+	pub fn div<N: TryInto<u128>>(lhs: N, rhs: FeeRate) -> rstd::result::Result<u128, &'static str> {
```

Comment:
`TryInto` is used instead of `Into` as the test for this uses `i32` for lhs:
```rust
assert_ok!(FeeRate::div(10, fee_rate), 9_u128)
```

Reply:
Since `safe_div` exists, let's just remove this and it's associated tests. It does the same thing now